### PR TITLE
gh-91996: Adding an HTTPMethod StrEnum to http

### PR DESCRIPTION
**Pull request by cibofo**
https://github.com/python/cpython/issues/91996

### CODE REVIEW

**Review by corona10 — CHANGES_REQUESTED** [review]
Please update the documentation if you want to add this enum

https://docs.python.org/3/library/http.html?highlight=httpstatus#http.HTTPStatus

File: Lib/http/__init__.py [hunk 1]
```diff
@@ -149,3 +149,30 @@ def __new__(cls, value, phrase, description=''):
...
+        * RFC 5789: PATCH Method for HTTP
+    """
+    def __new__(cls, value, description):
+        obj = str.__new__(cls, value)
+        obj._value_ = value
+
+        obj.description = description
+        return obj
+
+    GET = 'GET', 'Transfer a current representation of the target resource.'
```
Comment (davidism) [inline Lib/http/__init__.py]: These should be in some sort of order, probably alphabetical, possibly grouped by safety: `GET`, `HEAD`, `OPTIONS` / `DELETE`, `PATCH`, `POST`, `PUT` / `CONNECT`, `TRACE`.

**Comment by davidism** [comment]
While it's not technically a standard yet, would be good to be aware of https://www.ietf.org/archive/id/draft-ietf-httpbis-safe-method-w-body-02.html, the `QUERY` method to indicate a safe method (like `GET`) that has a body (like `POST`).

**Comment by davidism** [comment]
Also, consider the churn that will be introduced by this. I guarantee that some user of every library that uses HTTP will open issues asking for the library's methods to support these, which will then require every maintainer to evaluate whether they want to support it or keep rejecting requests to support it forever. Since all existing libraries use strings, they will have to add more complex code to accept both.

File: Lib/http/__init__.py (hunk 1 above)
Comment (cibofo) [reply Lib/http/__init__.py]: I used the same order from the RFCs, this is the usual order they are sorted (e.g. https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods)

**Comment by cibofo** [comment]
> While it's not technically a standard yet, would be good to be aware of https://www.ietf.org/archive/id/draft-ietf-httpbis-safe-method-w-body-02.html, the `QUERY` method to indicate a safe method (like `GET`) that has a body (like `POST`).
> 
I think it might be better to include the QUERY method in a seperate PR since it isn't part of the http standard and might change over time.

> This also misses any WebDAV methods, which users of Flask/Werkzeug occasionally ask about.
> 
Most of the users don't really know about webDAV and it might be confusing to them to see it, since HTTPMethod is such a small enum (unlike HTTPStatus) it is very noticeable.
Do you think that it would be better to add the webDAV methods?

**Comment by cibofo** [comment]
> Also, consider the churn that will be introduced by this. I guarantee that some user of every library that uses HTTP will open issues asking for the library's methods to support these, which will then require every maintainer to evaluate whether they want to support it or keep rejecting requests to support it forever. Since all existing libraries use strings, they will have to add more complex code to accept both.
>

I think that libraries that don't conform to the new higher standards shouldn't stop the improvement of cpython.
New programs and libraries being developed shouldn't miss out on new features just because already existing libraries don't use these features.
Cpython influences the entire python community and it should provide the best practices for creating new programs.

**Comment by cibofo** [comment]
Also, added documentation 
I have made the requested changes; please review again

**Comment by davidism** [comment]
> Most of the users don't really know about webDAV and it might be confusing to them to see it, since HTTPMethod is such a small enum (unlike HTTPStatus) it is very noticeable.

You've named the enum **HTTP*Method, but currently only represent the **REST* methods (and CONNECT and TRACE for some reason, even though they're probably even less common than WebDAV). From the RFC:

> Additional methods, outside the scope of this specification, have been standardized for use in HTTP. All such methods ought to be registered within the "Hypertext Transfer Protocol (HTTP) Method Registry" maintained by IANA, as defined in [Section 8.1](https://datatracker.ietf.org/doc/html/rfc7231#section-8.1).

https://www.iana.org/assignments/http-methods/http-methods.xhtml lists many more registered HTTP methods. Despite the RFC saying methods should be registered there, that's not a requirement, and in the end an HTTP method can be any string.

> I think that libraries that don't conform to the new higher standards shouldn't stop the improvement in cpython.

This isn't a "higher standard", it's just a different value for the same thing. This needs to be very clearly an improvement, not just a change.

**Comment by cibofo** [comment]
> You've named the enum _HTTP_Method, but currently only represent the _REST_ methods (and CONNECT and TRACE for some reason, even though they're probably even less common than WebDAV).
> 
Correct me if I am wrong but WebDAV is an extension to HTTP, not actual part of the standard, CONNECT and TRACE are.
Anyway, what WebDAV methods do you think should be added?

> This isn't a "higher standard", it's just a different value for the same thing. This needs to be very clearly an improvement, not just a change.
>
The code is more secure when using an enum, as magic strings are prone to mistypes and therefore errors, 
I (and many more) also think that they are more convenient than just using strings. 
So a more secure and convenient way to write code isn't a higher standard?

**Comment by davidism** [comment]
Convenient maybe. Secure no. Mistypes of http methods result in 405 errors, not security issues.

**Comment by cibofo** [comment]
> Convenient maybe, but it requires an import and a longer reference to type, compared to a string. Secure no. Mistypes of http methods result in 405 errors, not security issues.

Ok secure is not the right word, I meant less bugs

**Comment by cibofo** [comment]
I would appreciate your response

**Comment by davidism** [comment]
Sorry, I'm not a cpython maintainer. They're at a PyCon sprint right now so I'm sure their notifications are overflowing.

**Comment by ethanfurman** [comment]
@cibofo instead of merging `main` into your branch, rebase your branch.

**Comment by cibofo** [comment]
@ethanfurman  Done

**Review by ethanfurman — APPROVED** [review]

**Comment by ethanfurman** [comment]
@cibofo Please create a new PR that addes the other 31 or so registered methods.  Thanks!

**Comment by cibofo** [comment]
@ethanfurman You got it! gonna work on it tomorrow

**Comment by merwok** [comment]
> instead of merging main into your branch, rebase your branch.

That is contrary to usual workflow: https://devguide.python.org/pullrequest/

**Comment by ethanfurman** [comment]
> > instead of merging main into your branch, rebase your branch.
> 
> That is contrary to usual workflow: https://devguide.python.org/pullrequest/

Huh.  [Git Bootcamp](https://devguide.python.org/gitbootcamp/) uses `rebase`, although for older PRs.  The couple times I tried merging in `main` it was a disaster, but I was also learning how to use git and github, so it could have easily been something else.

I'll keep using `rebase` for my own work -- it makes me happy.  :-)

**Pull request merged**